storage/connectors/rediscluster/src/main/java/feast/storage/connectors/rediscluster/retriever/RedisClusterOnlineRetriever.java

```diff
@@ -0,0 +1,226 @@
+import feast.types.FieldProto.Field;
+import feast.types.ValueProto.Value;
+import io.grpc.Status;
+import io.lettuce.core.RedisURI;
+import io.lettuce.core.cluster.RedisClusterClient;
+import io.lettuce.core.cluster.api.StatefulRedisClusterConnection;
+import io.lettuce.core.cluster.api.sync.RedisAdvancedClusterCommands;
+import io.lettuce.core.codec.ByteArrayCodec;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+public class RedisClusterOnlineRetriever implements OnlineRetriever {
+
+  private final RedisAdvancedClusterCommands<byte[], byte[]> syncCommands;
+
+  private RedisClusterOnlineRetriever(StatefulRedisClusterConnection<byte[], byte[]> connection) {
+    this.syncCommands = connection.sync();
+  }
+
+  public static OnlineRetriever create(Map<String, String> config) {
+    List<RedisURI> redisURIList =
+        Arrays.stream(config.get("connection_string").split(","))
+            .map(
+                hostPort -> {
+                  String[] hostPortSplit = hostPort.trim().split(":");
+                  return RedisURI.create(hostPortSplit[0], Integer.parseInt(hostPortSplit[1]));
+                })
+            .collect(Collectors.toList());
+
+    StatefulRedisClusterConnection<byte[], byte[]> connection =
+        RedisClusterClient.create(redisURIList).connect(new ByteArrayCodec());
+
+    return new RedisClusterOnlineRetriever(connection);
+  }
+
+  public static OnlineRetriever create(StatefulRedisClusterConnection<byte[], byte[]> connection) {
+    return new RedisClusterOnlineRetriever(connection);
+  }
+
+  /**
+   * Gets online features from redis. This method returns a list of {@link FeatureRow}s
+   * corresponding to each feature set spec. Each feature row in the list then corresponds to an
+   * {@link EntityRow} provided by the user.
+   *
+   * @param entityRows list of entity rows in the feature request
+   * @param featureSetRequests Map of {@link FeatureSetSpec} to feature references in the request
+   *     tied to that feature set.
+   * @return List of List of {@link FeatureRow}
+   */
+  @Override
+  public List<List<FeatureRow>> getOnlineFeatures(
+      List<EntityRow> entityRows, List<FeatureSetRequest> featureSetRequests) {
+
+    List<List<FeatureRow>> featureRows = new ArrayList<>();
+    for (FeatureSetRequest featureSetRequest : featureSetRequests) {
+      List<RedisKey> redisKeys = buildRedisKeys(entityRows, featureSetRequest.getSpec());
+      try {
+        List<FeatureRow> featureRowsForFeatureSet =
+            sendAndProcessMultiGet(
+                redisKeys,
+                featureSetRequest.getSpec(),
+                featureSetRequest.getFeatureReferences().asList());
+        featureRows.add(featureRowsForFeatureSet);
+      } catch (InvalidProtocolBufferException | ExecutionException e) {
+        throw Status.INTERNAL
```

Comment:
Just wondering if we should be throwing gRPC exceptions so deep within the code base. Perhaps this should be left to higher layers?

Reply:
While i agree that this should be left to higher layers, it is better to implemented that in a separate PR, as RedisOnlineRetriever will also be affected by the change.